### PR TITLE
feat(BBD 748): SAYT to include Past Purchase selections

### PR DESCRIPTION
**Pull request by Filip-F**
Depends on https://github.com/groupby/flux-capacitor/pull/120

### CODE REVIEW

File: src/autocomplete/index.ts [hunk 1]
```diff
@@ -88,11 +91,15 @@ class Autocomplete {
     }
   }
 
-  updateProducts({ dataset: { query: selectedQuery, refinement, field } }: HTMLElement) {
+  updateProducts({ dataset: { query: selectedQuery, refinement, field, pastPurchase } }: HTMLElement) {
     const query = selectedQuery == null ? this.select(Selectors.autocompleteQuery) : selectedQuery;
     this.flux.emit('query:update', query);
-    // tslint:disable-next-line max-line-length
-    this.flux.saytProducts(field ? null : query, refinement ? [{ field: field || this.state.category, value: refinement }] : []);
+    if (pastPurchase !== undefined) {
```
Comment (effervescentia) [inline src/autocomplete/index.ts:97]: will this just be truthy when we want to make a `pastPurchase` request?

File: src/past-purchase/index.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { alias, tag, Events, Selectors, Store, Tag } from '@storefront/core';
+
+@alias('saytCategories')
```
Comment (effervescentia) [inline src/past-purchase/index.ts]: do we need this alias for the tag's state?

File: src/past-purchase/index.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+@tag('gb-sayt-past-purchase', require('./index.html'))
+class PastPurchase {
+  $pastPurchase: Store.Autocomplete.Suggestion;
+
+  state: PastPurchase.State = {
+    onClick: (event: MouseEvent) => {
+      // todo:
```
Comment (effervescentia) [inline src/past-purchase/index.ts:9]: work pending?

File: test/unit/autocomplete.ts [hunk 4]
```diff
@@ -38,6 +38,44 @@ suite('Autocomplete', ({ expect, spy, stub }) => {
           navigations: NAVIGATIONS
         });
       });
+
+      describe('onHover()', () => {
+        const targets = [0,'tar',2,3,4];
```
Comment (effervescentia) [inline test/unit/autocomplete.ts:43]: add newline between `const` and `beforeEach()`

File: test/unit/autocomplete.ts [hunk 5]
```diff
@@ -38,6 +38,44 @@ suite('Autocomplete', ({ expect, spy, stub }) => {
...
+
+        it('should not call setActivation with false activate if this.isActive() is false', () => {
+          const target = 'tar';
+
+          autocomplete.state.selected = 0;
+          autocomplete.isActive = spy(() => false);
+          autocomplete.state.onHover(<any>{ target });
+
+          expect(autocomplete.setActivation).to.be.calledOnce
+            .and.calledWith(targets, Array.from(targets).findIndex((element) => element === target), true);
```
Comment (effervescentia) [inline test/unit/autocomplete.ts]: why do we need to call `Array.from()` on `targets`

Comment (effervescentia) [reply test/unit/autocomplete.ts]: and shouldn't we just call `indexOf()` to get the index of `target`

File: test/unit/autocomplete.ts [hunk 6]
```diff
@@ -38,6 +38,44 @@ suite('Autocomplete', ({ expect, spy, stub }) => {
...
+        it('should call setActivation with false activate if this.isActive() is true', () => {
+          const target = 'tar';
+
+          autocomplete.state.selected = 0;
+          autocomplete.isActive = spy(() => true);
+          autocomplete.state.onHover(<any>{ target });
+
+          expect(autocomplete.setActivation).to.be.calledTwice
+            .and.calledWith(targets, autocomplete.state.selected, false)
+            .and.calledWith(targets, Array.from(targets).findIndex((element) => element === target), true);
```
Comment (effervescentia) [inline test/unit/autocomplete.ts]: `Array.from().findIndex()` to `[].indexOf()`

File: src/autocomplete/index.ts [hunk 7]
```diff
@@ -8,6 +8,9 @@ class Autocomplete {
   state: Autocomplete.State = <any>{
     onHover: (event: MouseEvent) => {
       const targets = this.activationTargets();
+      if (Array.from(targets).findIndex((element) => element === event.target) === this.state.selected) {
```
Comment (effervescentia) [inline src/autocomplete/index.ts]: `Array.from().findIndex()` vs `[].indexOf()`?

File: test/unit/past-purchase.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import suite from './_suite';
+
+const PAST = ['a', 'b'];
+
+suite('PastPurchase', ({ expect, spy, stub }) => {
+  let pastPurchase: PastPurchase;
+  let queryPastPurchases: sinon.SinonStub;
+
+  beforeEach(() => {
+    PastPurchase.prototype.select = <any>spy(() => PAST);
```
Comment (effervescentia) [inline test/unit/past-purchase.ts:13]: if we modify the `prototype` here then we should be deleting the added key in an `afterEach()`

**Review by effervescentia — CHANGES_REQUESTED** [review]
looks like there may be some changes still pending
probably needs a rebase and to target the latest core version
`yarn add -P @storefront/core@~X.X.X`

File: src/autocomplete/index.ts (hunk 1 above)
Comment (Filip-F) [reply src/autocomplete/index.ts:97]: I could check for truthiness but it was just to have it so passing in anything aside from undefined registers it as passed in.

File: src/past-purchase/index.ts (hunk 3 above)
Comment (Filip-F) [reply src/past-purchase/index.ts:9]: Yes, the routing wasn't done at this time, that's part of the next block of work on BBD-833. Would you prefer I remove it entirely in this ticket?

**Pull request closed without merging**